File: library/src/main/java/com/medavox/library/mutime/ParallelProcess.java
package com.medavox.library.mutime;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by scc on 21/10/17.
 */
class ParallelProcess<In, Out> extends Thread {
    private int numThreads = -1;
    private List<InternalWrapper> threads;
    private In inputSingle = null;
    private In[] inputArray = null;
    private boolean arrayInputMode;

    public ParallelProcess(In[] input) {
        arrayInputMode = true;
        this.inputArray = input;
        numThreads = input.length;
    }

    public ParallelProcess(In input, int numberOfThreads) {
        this.inputSingle = input;
        arrayInputMode = false;
        numThreads = numberOfThreads;
    }


    public void doWork(Worker<In, Out> smug) {
        threads = new ArrayList<InternalWrapper>(numThreads);
        if(arrayInputMode) {
            for (int i = 0; i < inputArray.length; i++) {
                InternalWrapper iw = new InternalWrapper(inputArray[i], smug);
                threads.add(i, iw);
                iw.start();
            }
        }
        else {
            for (int i = 0; i < numThreads; i++) {
                InternalWrapper iw = new InternalWrapper(inputSingle, smug);
                threads.add(i, iw);
                iw.start();
            }
        }
    }

    public void collectOutputWhenFinished(Out[] output) {
        if(output.length != numThreads) {
            throw new IllegalArgumentException("Supplied output array length must match number of" +
                " threads. Number of threads: "+numThreads+"; output array length: "+output.length);
        }
        for (int i = 0; i < threads.size(); i++) {
            try {
                InternalWrapper t = threads.get(i);
                if(t != null) {
                    t.join();
                    output[i] = t.getOutput();
                }
            } catch (InterruptedException ie) {
                System.err.println("" + ie);
                ie.printStackTrace();
            }
        }
    }

    private class InternalWrapper extends Thread {
        private In in;
        private Out out;
        private Worker<In, Out> worker;
        public InternalWrapper(In inn, Worker<In, Out> worker) {
            in = inn;
            this.worker = worker;
        }
        @Override
        public void run() {
            super.run();
            out = worker.performWork(in);
        }

        public Out getOutput() {
            return out;
        }
    }

    public interface Worker<In, Out> {
        Out performWork(In input);
    }
}

File: library/src/main/java/com/medavox/library/mutime/ParallelProcess.kt
package com.medavox.library.mutime

/**Performs operations on input data concurrently.
 *
 * @author Adam Howard
 * @since 17/09/2018
 */
internal class ParallelProcess<In, Out>(private val worker:(input:In) -> Out) : Thread() {
    private val threads:MutableSet<InternalWrapper> = mutableSetOf()

    /**Run this instance's worker repeatedly concurrently with the same input.
     * @param numberOfThreads the number of threads --
     * and the number of times to repeat -- running the worker
     * @param input the input to run on*/
    fun repeatOnInput(numberOfThreads:Int, input:In) {
        for(i in 1..numberOfThreads) {
            val worker = InternalWrapper(input, worker)
            threads.add(worker)
            worker.start()
        }
    }

    /**Process an array in parallel, with one worker thread per array element.
     * @param input the array to iterate over concurrently.*/
    fun oneWorkerPerElement(input:Array<out In>) {
        for(i in input.indices) {
            val worker = InternalWrapper(input[i], worker)
            threads.add(worker)
            worker.start()
        }
    }

    /**Get the resulting output from the worker threads.
     * This is a blocking method; it only returns once all worker threads have finished.*/
    fun collectOutputWhenFinished(): List<Out> {
        val outputInProgress = mutableSetOf<Out?>()//the worker may fail to produce output
        for(thread in threads) {
            thread.join()
            outputInProgress.add(thread.getOutput())
        }
        //get rid of null elements, convert from MutableSet<Out?> to List<Out>
        return outputInProgress.asSequence().filter{it != null}.map{it as Out}.toList()
    }

    private inner class InternalWrapper(val input: In, val workFunction:(input:In) -> Out): Thread() {
        private var out: Out? = null

        override fun run() {
            super.run()
            out = workFunction(input)
        }

        fun getOutput(): Out? {
            return out
        }
    }
}